contrib/win32/win32compat/tncon.c

```diff
@@ -188,7 +188,12 @@ ReadConsoleForTermEmul(HANDLE hInput, char *destin, int destinlen)
 					}
 
 					if (isConsoleVTSeqAvailable) {
-						if (inputRecord.Event.KeyEvent.uChar.UnicodeChar != L'\0') {
+						// Ctrl+Space & Ctrl+@ generate a NULL character but should still be sent
+						DWORD dwControlKeyState = inputRecord.Event.KeyEvent.dwControlKeyState;
+						DWORD dwCtrlPressed = (dwControlKeyState & (LEFT_CTRL_PRESSED | RIGHT_CTRL_PRESSED));
```

Comment:
Only thing I'm concerned about is the `!IsAltPressed()` check from the terminal side. Presumably that was there for a reason. @zadjii-msft? Do you recall?

My theory is that it had something to do with <kbd>AltGr</kbd> as that can be simulated with a <kbd>Ctrl</kbd> and an <kbd>Alt</kbd> (https://devblogs.microsoft.com/oldnewthing/20040329-00/?p=40003#:~:text=The%20combination%20Ctrl%2BAlt%20is%20also%20known%20as%20AltGr%2C,keyboards%20there%20are%20only%20two%20%28Normal%20and%20Shift%29.)

Reply:
(@lhecker might also know as he was in AltGr land for a while)

Reply:
Yes I've done AltGr handling once.

![meme](https://user-images.githubusercontent.com/2256941/155432414-877057f9-e53b-45bc-b3a3-18f2111a5d1e.png)

Reply:
The change appears to be correct from what I can tell.
The code that is relevant here are these two places:
* https://github.com/microsoft/terminal/blob/9ab4abfb5471e91235e34379bf90e24ec3a22c67/src/terminal/input/terminalInput.cpp#L594-L601
* https://github.com/microsoft/terminal/blob/9ab4abfb5471e91235e34379bf90e24ec3a22c67/src/terminal/input/terminalInput.cpp#L635-L642

---

However I believe - maybe? - that this PR is too complex.
From what I can tell the non-VT codepath performs the exact same conversion using `WideCharToMultiByte`, followed by a large switch-case. In the default branch, if conversion fails, it'll call `WriteToBuffer((char *)octets, n);`.

So this leads me to believe that the proper fix here is to simply remove the `if (inputRecord.Event.KeyEvent.uChar.UnicodeChar != L'\0')` condition entirely, because that would mirror the non-VT codepath.

Reply:
To be more specific, I believe a solution might look like this (includes Ctrl+Space support):
```c
// Returns true if this is a Ctrl+Space or Ctrl+@ combination.
// It assumes that inputRecord->Event.KeyEvent.uChar.UnicodeChar is 0.
static bool
IsNullKeySequence(const INPUT_RECORD* inputRecord)
{
    DWORD state = inputRecord->Event.KeyEvent.dwControlKeyState;
    WORD code = inputRecord->Event.KeyEvent.wVirtualKeyCode;
    return (state & (LEFT_CTRL_PRESSED | RIGHT_CTRL_PRESSED))
        && (code == LOBYTE(VkKeyScanW(0)) || code == LOBYTE(VkKeyScanW(L' ')));
}
```

And then use it as:
```c
if (inputRecord.Event.KeyEvent.uChar.UnicodeChar != L'\0' || IsNullKeySequence(&inputRecord)) {
```

* I'm abusing C's support for "[short-circuit evaluation](https://en.wikipedia.org/wiki/Short-circuit_evaluation)"
  By putting function calls into the right-hand-side of boolean conditions we can make sure that they're only called if the other side isn't already "satisfied". Basically with this in place, `IsNullKeySequence` will only be called if `UnicodeChar` is actually 0.
* I'm passing `INPUT_RECORD` by pointer ("by reference" in other languages), so that we don't incur a copy. `GetVTSeqFromKeyStroke` should do the same.
* `static` because the function is local to this file
* If this code doesn't compile because of the "bool", replace it with `_Bool`, which is an official alias and will always work.

This code still doesn't support proper Alt handling, but I'm not sure how important that is for an initial bug fix. The previous code seemingly doesn't support it either after all. 🙂

Reply:
Ahhh, got it. Thanks for the detailed response!